agent/xds/clusters.go

```diff
@@ -159,25 +159,35 @@ func (s *Server) clustersFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsho
 	// generate the per-service clusters
 	for svc, _ := range cfgSnap.MeshGateway.ServiceGroups {
 		clusterName := connect.ServiceSNI(svc.ID, "", svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+		resolver, hasResolver := cfgSnap.MeshGateway.ServiceResolvers[svc]
 
-		cluster, err := s.makeMeshGatewayCluster(clusterName, cfgSnap)
+		// Create the cluster using mesh gateway configs for default/unnamed services
+		// without a service-router
+		if !hasResolver {
+			cluster, err := s.makeMeshGatewayCluster(clusterName, cfgSnap)
+			if err != nil {
+				return nil, err
+			}
+			clusters = append(clusters, cluster)
+			continue
```

Comment:
It was immediately clear to me why there was a `continue` here. I eventually figured it out that it was to avoid the code below when we already know we don't have a resolver (to prevent some segfaults)


Could there be a way to refactor this code a little to make the code flow a little clearer? Maybe a way that we only generate the unnamed subsets cluster in 1 location?

Reply:
Updated! Happy to improve readability